### ISSUE TITLE
Feature: suppress/disable rules in source

### ISSUE THREAD

**Issue by bgotink** [SonarSource/SonarTS#680]
**I want to request a feature.**

Sometimes rules are overzealous. In Java we can ignore rules within a certain scope using the standard `@SuppressWarnings`, it would be nice if we could ignore rules in sonarts as well, e.g. using the tslint standard `// tslint:disable-next-line` or `// tslint:disable`.

I know SonarQube allows marking issues as false positives, but we prefer marking false positives in code instead of on an external tool. That makes it easier to manage them and it keeps working throughout refactors whereas those issues tend to pop back up in SonarQube once in a blue moon.

Example scenario: bitwise operators are disallowed, but for a certain calculation we need them so we want to grant an exception, documenting them in code. Example usage:

```ts
// actionThatMightThrow might throw an error, so replacing the line with
//   Promise.resolve(someLib.actionThatMightThrow())
// would result in an error being thrown rather than the rejected promise we now get
// tslint:disable-next-line:prefer-promise-shorthand
const somePromise = new Promise(resolve => resolve(someLib.actionThatMightThrow()));
```

Sidenote: we use tslint for local development, using `"extends": ["tslint-sonarts"],` in our `tslint.json`. Disabling the rules works fine for local development, but SonarQube doesn't listen to the tslint:disable directives.

**Comment by vilchik-elena** [comment]
Hey,

I think for the use case like yours, the best way to go is to use external issues (see [doc](https://docs.sonarqube.org/display/PLUG/Importing+TSLint+and+ESLint+issues)). Create an empty rules profile for TS in your SQ to avoid issue duplicates (even if we have reduplication logic it will not help when issue is suppressed with tslint comment). 

This way you will have in SQ exactly the same issues as locally.

Please let us know what do you think about it:)

**Comment by bgotink** [comment]
That solution is at the moment not technically feasible for us as we won't update SonarQube to a version that allows external issues before start of next year.

Regardless of our current technical issues, we fear that maintaining configuration in two locations would make it hard to properly configure the sonar scanner. We'd still need some configuration in the profile, e.g. coverage rules.

**Comment by vilchik-elena** [comment]
I see your point, but I think it's worth trying

**Comment by saberduck** [comment]
@realshaft we simply don't have bandwidth to do everything (even previous docker images were unofficial). You can vote for the new feature here https://community.sonarsource.com/t/docker-container-for-7-3/1851

**Comment by lhude** [comment]
I also think that it would be a great improvment to be able to disable rules in source code, for 2 reasons :
- as @bgotink said, the issues marked in the UI may reappear for some reason after refactors
- it doesn't work accross branches, you have to mark the issues on each branch

The second point it is even mentionned in the [SonarQube documentation](https://docs.sonarqube.org/display/SONAR/Frequently+Asked+Questions#FrequentlyAskedQuestions-False-PositiveandWon'tFix) : 

> You can mark individual issues as False Positive or Won't Fix through the issues interface. However, this solution doesn't work across branches - you'll have to re-mark the issue False Positive for each branch under analysis. So an in-code approach may be preferable if multiple branches of a project are under analysis

I'm aware of the "//NOSONAR" possibility, but it is too "agressive" to disable all the rules on the line you put this comment on, and it is thereby considered as a bad practice. 
Plus it doesn't seem to work anymore.

It would be great to have the possibility to disable a rule, based on its rulekey.
Could you please consider to reopen this issue ?

**Comment by vilchik-elena** [comment]
@lhude I would say this issue is more related to entire Sonar ecosystem, not only SonarJS analyzer. I suggest you to create a thread on https://community.sonarsource.org/
